### PR TITLE
Branchstat algorithm

### PR DESCRIPTION
**Pull request by jeromekelleher**
Refactor of the branch/node stat algorithms. Changes

1. Store the value of f(x) during the branch_stat algorithm, which simplifies the logic a bit and reduces the number of calls of f substantially. (not fully implemented in C when the PR was opened)
2. Abstract out the details of running polarised vs unpolarised summary functions in C for node and branch stats. (It would be pleasing if this was possible for site stats also, but I don't see how as polarisation is much more central to how things are actually computed.)

Should be significant perf gains from this. I'll do some before/after comparisons when its done.

### CODE REVIEW

**Comment by jeromekelleher** [comment]
This can close #257 I think --- getting the function to update now would be very minor perf tweak, probably not worth it. I haven't thought about whether this idea lets us abstract to the AFS yet. That would be nice.

**Comment by jeromekelleher** [comment]
Benchmarks:

On this tree sequence:
```
$ python3 -m tskit info large-example.trees 
sequence_length:  100000000.0
trees:            456789
samples:          100000
individuals:      0
nodes:            500814
edges:            1941686
sites:            484395
mutations:        484395
migrations:       0
populations:      1
provenances:      1
```
and this code:
```python
import tskit
import time
import numpy as np

ts = tskit.load("large-example.trees")


print("mode", "n_sets", "time", sep="\t")
for mode in ["site", "branch", "node"]:
    for num_sample_sets in [1, 10, 100]:
        sample_sets = np.array_split(ts.samples(), num_sample_sets)
        before = time.perf_counter()
        x = ts.diversity(sample_sets=sample_sets, mode=mode)
        duration = time.perf_counter() - before
        print(mode, num_sample_sets, "{:.3f}".format(duration), sep="\t")
```

On upstream/master we get:
```
$ python3 benchmark.py
mode    n_sets  time
site    1       0.344
site    10      0.557
site    100     2.413
branch  1       1.767
branch  10      2.949
branch  100     15.900
node    1       1.111
node    10      2.162
node    100     12.033
```
On this branch we get 
```
$ python3 benchmark.py
mode    n_sets  time
site    1       0.349
site    10      0.557
site    100     2.426
branch  1       0.647
branch  10      1.761
branch  100     11.555
node    1       0.570
node    10      1.654
node    100     10.541
```

So, we do get a decent perf jump on branch stats when we have a small number of sample sets. Not sure what's going on with the larger number of sample sets, will investigate.

**Comment by jeromekelleher** [comment]
I don't think we're going to do much better here. When we have 100 sample sets, we get this from perf:
```
  44.93%  python3        _tskit.cpython-36m-x86_64-linux-gnu.so             [.] tsk_treeseq_general_stat                                                                                                                                      
  36.67%  python3        _tskit.cpython-36m-x86_64-linux-gnu.so             [.] diversity_summary_func                                                                                                                                        
  14.79%  python3        _tskit.cpython-36m-x86_64-linux-gnu.so             [.] npolarised_summary_func                                                                                                                                      
   0.65%  python3        [kernel]                                           [k] 0xffffffffb7c009e7                                                                                                                                            
   0.45%  python3        _tskit.cpython-36m-x86_64-linux-gnu.so             [.] tsk_treeseq_init                                                                                                                                              
   0.26%  python3        python3.6                                          [.] _PyEval_EvalFrameDefault           
```
When you dig in we're spending time in looping over the sample sets, as you might expect. We could probably squeeze a bit more perf here by using ``restrict``s, but they are tricky and I don't think we'll get much out of it.

The reason that node and branch stats are so much slower than site stats is because they're doing a lot more. The summary func is only called on mutation nodes not on *every* node. So, it's as good as it's going to get for now I think, and we should be pretty happy with this (it's friggin fast!).

**Comment by jeromekelleher** [comment]
@petrelharp, any thoughts here? I'm going to merge this and write up the algorithm unless you object.

**Comment by petrelharp** [comment]
Nice performance improvements!

> Abstract out the details of running polarised vs unpolarised summary functions in C for node and branch stats. (It would be pleasing if this was possible for site stats also, but I don't see how as polarisation is much more central to how things are actually computed.)

This is a nice way to do it. Doing it the same way for site stats would require passing a list of states (one for each derived allele) to a wrapped function, I think - we could work this out but it doesn't seem necessary.

File: c/tskit/trees.c [hunk 1]
```diff
@@ -973,6 +973,8 @@ tsk_treeseq_check_windows(tsk_treeseq_t *self, size_t num_windows, double *windo
     return ret;
 }
 
+/* TODO make these functions more consistent in how the arguments are ordered */
```
Comment (petrelharp) [inline c/tskit/trees.c]: :+1:

**Comment by petrelharp** [comment]
This is very nice and clean. I like it.

**Pull request merged**